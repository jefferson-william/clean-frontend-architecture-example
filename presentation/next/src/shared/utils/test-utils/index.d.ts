import { Queries, RenderResult } from '@testing-library/react'
import { Theme } from '@material-ui/core'
import { Router } from 'next/router'
import States from '~/types/store/rootStates'
import customQueries, { MyQueries } from '~/shared/utils/test-utils/custom-queries'

export interface MyRenderOptions {
  router?: Partial<Router>
  theme?: Theme
  initialState?: States
  store?(states?: States): Store<any, any>
}

export interface MyRenderedOptions {
  router?: Partial<Router>
  theme?: Theme
  initialState?: States
  store: Store<any, any>
}
