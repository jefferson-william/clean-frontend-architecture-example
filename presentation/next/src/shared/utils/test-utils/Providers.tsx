import { Provider } from 'react-redux'
import { RouterContext } from 'next-server/dist/lib/router-context'
import { ThemeProvider } from 'styled-components'
import { Store } from 'redux'
import { MyRenderOptions } from '~/shared/utils/test-utils/index.d'

const Providers = (myStore: Store<any, any>, allOptions: MyRenderOptions) => ({ children }: any) => (
  <RouterContext.Provider value={allOptions.router as any}>
    <Provider store={myStore}>
      <ThemeProvider theme={allOptions.theme}>{children}</ThemeProvider>
    </Provider>
  </RouterContext.Provider>
)

export default Providers
