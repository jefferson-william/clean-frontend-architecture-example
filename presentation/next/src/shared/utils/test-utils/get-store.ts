import { Store } from 'redux'
import { store } from '~/shared/store'
import rootStates from '~/shared/store/rootStates'
import { MyRenderOptions } from '~/shared/utils/test-utils/index.d'

const getStore = (options: MyRenderOptions): Store<any, any> => {
  const state = options.initialState || rootStates

  return (options?.store && options!.store(state)) || store(state)
}

export default getStore
