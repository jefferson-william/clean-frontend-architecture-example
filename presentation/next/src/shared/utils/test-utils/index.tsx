import { render, waitFor, queries, RenderResult } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import theme from '~/shared/styles/theme'
import getStore from '~/shared/utils/test-utils/get-store'
import { MyRenderOptions, MyRenderedOptions } from '~/shared/utils/test-utils/index.d'
import Providers from '~/shared/utils/test-utils/Providers'
import customQueries, { MyQueries } from './custom-queries'

export const myRender = async (
  Component: any,
  options?: MyRenderOptions
): Promise<[MyRenderResult, MyRenderedOptions]> => {
  const allOptions = {
    ...options,
  }

  allOptions.theme = options?.theme || theme

  const myStore = getStore(allOptions)

  const rendered = render<MyQueries, HTMLElement>(Component, {
    wrapper: Providers(myStore, allOptions),
    queries: { ...queries, ...customQueries } as MyQueries,
  })

  await waitFor(() => rendered.container.querySelector('body'))

  return [rendered, { store: myStore, ...allOptions }]
}

export * from '@testing-library/react'

export type MyRenderResult = RenderResult<MyQueries, HTMLElement>

export { myRender as render, userEvent }
