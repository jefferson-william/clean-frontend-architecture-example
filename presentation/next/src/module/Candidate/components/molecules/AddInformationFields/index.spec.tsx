import {
  populateWhereDidYouWorkForm,
  testAddAndRemoveFieldsOfWhereDidYouWorkForm,
} from '~/module/Candidate/components/molecules/AddInformationFields/actions'
import AddInformationFieldsStub from '~/module/Candidate/components/molecules/AddInformationFields/stub'
import getFormData from '~/shared/utils/form/get-form-data'
import { render, act, MyRenderResult } from '~/shared/utils/test-utils'

describe('components/AddInformationFields', () => {
  let wrapper: MyRenderResult

  describe('when not have default params', () => {
    beforeEach(async () => {
      ;[wrapper] = await render(<AddInformationFieldsStub list={['']} />)
    })

    it('should to match snapshot', () => {
      expect(wrapper.asFragment()).toMatchSnapshot()
    })

    describe('when test behavior', () => {
      it('all interactions should work', async () => {
        await act(async () => {
          await testAddAndRemoveFieldsOfWhereDidYouWorkForm(wrapper)
        })
      })
    })

    describe('when fill in all fields correctly', () => {
      it('should return all the values filled in', async () => {
        await act(async () => {
          await populateWhereDidYouWorkForm(wrapper)

          expect(getFormData(wrapper.container.querySelector('form') as HTMLFormElement)).toMatchObject({
            'whereDidYouWork[0]': 'Amazon',
            'whereDidYouWork[1]': 'Yahoo',
          })
        })
      })
    })
  })

  describe('when have two default params', () => {
    beforeEach(async () => {
      ;[wrapper] = await render(<AddInformationFieldsStub list={['Google', 'Microsoft']} />)
    })

    it('should to match snapshot', () => {
      expect(wrapper.asFragment()).toMatchSnapshot()
    })
  })
})
