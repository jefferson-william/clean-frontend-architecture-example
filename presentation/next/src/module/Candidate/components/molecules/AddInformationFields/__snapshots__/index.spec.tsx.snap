// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/AddInformationFields when have two default params should to match snapshot 1`] = `
<DocumentFragment>
  <form>
    <div
      class="MuiFormControl-root main__form-control"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled Mui-required Mui-required"
        data-shrink="true"
        for="whereDidYouWork0"
      >
        Onde já trabalhou?
        <span
          aria-hidden="true"
          class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
        >
           *
        </span>
      </label>
      <div
        class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl MuiInputBase-adornedEnd"
      >
        <input
          aria-invalid="false"
          class="MuiInputBase-input MuiInput-input MuiInputBase-inputAdornedEnd"
          id="whereDidYouWork0"
          name="whereDidYouWork[0]"
          required=""
          type="text"
          value="Google"
        />
        <div
          class="MuiInputAdornment-root MuiInputAdornment-positionEnd"
        >
          <button
            aria-label="Remover"
            class="MuiButtonBase-root MuiIconButton-root"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiIconButton-label"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                />
              </svg>
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </div>
    <div
      class="MuiFormControl-root main__form-control"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled Mui-required Mui-required"
        data-shrink="true"
        for="whereDidYouWork1"
      >
        Onde já trabalhou?
        <span
          aria-hidden="true"
          class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
        >
           *
        </span>
      </label>
      <div
        class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl MuiInputBase-adornedEnd"
      >
        <input
          aria-invalid="false"
          class="MuiInputBase-input MuiInput-input MuiInputBase-inputAdornedEnd"
          id="whereDidYouWork1"
          name="whereDidYouWork[1]"
          required=""
          type="text"
          value="Microsoft"
        />
        <div
          class="MuiInputAdornment-root MuiInputAdornment-positionEnd"
        >
          <button
            aria-label="Remover"
            class="MuiButtonBase-root MuiIconButton-root"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiIconButton-label"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                />
              </svg>
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
          <button
            aria-label="Adicionar"
            class="MuiButtonBase-root MuiIconButton-root"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiIconButton-label"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
                />
              </svg>
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </div>
  </form>
</DocumentFragment>
`;

exports[`components/AddInformationFields when not have default params should to match snapshot 1`] = `
<DocumentFragment>
  <form>
    <div
      class="MuiFormControl-root main__form-control"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink Mui-required Mui-required"
        data-shrink="true"
        for="whereDidYouWork0"
      >
        Onde já trabalhou?
        <span
          aria-hidden="true"
          class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
        >
           *
        </span>
      </label>
      <div
        class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl MuiInputBase-adornedEnd"
      >
        <input
          aria-invalid="false"
          class="MuiInputBase-input MuiInput-input MuiInputBase-inputAdornedEnd"
          id="whereDidYouWork0"
          name="whereDidYouWork[0]"
          required=""
          type="text"
          value=""
        />
        <div
          class="MuiInputAdornment-root MuiInputAdornment-positionEnd"
        >
          <button
            aria-label="Remover"
            class="MuiButtonBase-root MuiIconButton-root Mui-disabled Mui-disabled"
            disabled=""
            tabindex="-1"
            type="button"
          >
            <span
              class="MuiIconButton-label"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                />
              </svg>
            </span>
          </button>
          <button
            aria-label="Adicionar"
            class="MuiButtonBase-root MuiIconButton-root"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiIconButton-label"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
                />
              </svg>
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </div>
  </form>
</DocumentFragment>
`;
